src/js/node/http2.ts

```diff
@@ -248,31 +351,920 @@ const constants = {
+      socket = options.createConnection(url, options);
+      this[bunHTTP2Socket] = socket;
+      if (socket.secureConnecting === true) {
+        socket.on("secureConnect", this.#onConnect.bind(this));
+      } else if (socket.connecting === true) {
+        socket.on("connect", this.#onConnect.bind(this));
+      } else {
+        process.nextTick(this.#onConnect.bind(this));
+      }
+    } else {
+      socket = connectWithProtocol(
+        protocol,
+        options
+          ? {
+              host: url.hostname,
+              port,
+              ALPNProtocols: ["h2", "http/1.1"],
+              ...options,
+            }
+          : {
+              host: url.hostname,
+              port,
+              ALPNProtocols: ["h2", "http/1.1"],
+            },
+        this.#onConnect.bind(this),
+      );
+      this[bunHTTP2Socket] = socket;
+    }
+    this.#encrypted = socket instanceof TLSSocket;
+
+    this.#parser = new H2FrameParser({
+      context: this,
+      settings: options,
+      handlers: ClientHttp2Session.#Handlers,
+    });
+
+    socket.on("close", this.#onClose.bind(this));
+    socket.on("error", this.#onError.bind(this));
+    socket.on("timeout", this.#onTimeout.bind(this));
+  }
+
+  // Gracefully closes the Http2Session, allowing any existing streams to complete on their own and preventing new Http2Stream instances from being created. Once closed, http2session.destroy() might be called if there are no open Http2Stream instances.
+  // If specified, the callback function is registered as a handler for the 'close' event.
+  close(callback: Function) {
+    this.#closed = true;
+
+    if (typeof callback === "function") {
+      this.once("close", callback);
+    }
+    if (this.#connections === 0) {
+      this.destroy();
+    }
+  }
+
+  destroy(error?: Error, code?: number) {
+    const socket = this[bunHTTP2Socket];
+    this.#closed = true;
+    code = code || constants.NGHTTP2_NO_ERROR;
+    if (socket) {
+      this.goaway(code, 0, Buffer.alloc(0));
+      socket.end();
+    }
+    this[bunHTTP2Socket] = null;
+    // this should not be needed since RST + GOAWAY should be sent
+    for (let [_, stream] of this.#streams) {
+      if (error) {
+        stream.emit("error", error);
+      }
+      stream.destroy();
+      stream.rstCode = code;
+      stream.emit("close");
+    }
+
+    if (error) {
+      this.emit("error", error);
+    }
+
+    this.emit("close");
+  }
+
+  request(headers: any, options?: any) {
+    if (this.destroyed || this.closed) {
+      const error = new Error(`ERR_HTTP2_INVALID_STREAM: The stream has been destroyed`);
+      error.code = "ERR_HTTP2_INVALID_STREAM";
+      throw error;
+    }
+
+    if (this.sentTrailers) {
+      const error = new Error(`ERR_HTTP2_TRAILERS_ALREADY_SENT: Trailing headers have already been sent`);
+      error.code = "ERR_HTTP2_TRAILERS_ALREADY_SENT";
+      throw error;
+    }
+
+    if (!(headers instanceof Object)) {
```

Comment:
```suggestion
    if (!($isObject(headers)) {
```

Reply:
fixed in https://github.com/oven-sh/bun/pull/6761/commits/946a993adea0d7084e96012fd60337da5be5bc49